### PR TITLE
chore: Rework Simulations scan

### PR DESCRIPTION
**Pull request by slandelle**
Motivation:

The existing file based filter is bad, in particular as it requires naming the Simulation classes as `XXXSimulation`.

Modification:

* use standard scan mechanism
* don't use filters for the user defined simulation class name (gatlingRun-FQCN)
* replace file path filters matching source files with ant based filters on class names

### CODE REVIEW

File: src/test/groovy/func/GatlingRunTaskTest.groovy [hunk 1]
```diff
@@ -32,9 +33,9 @@ class GatlingRunTaskTest extends GatlingUnitSpec {
 
     def "should resolve simulations using custom filter"() {
         given:
-        project.gatling.simulations = { include "**/*AdvancedSimulation*" }
+        project.gatling.includes = [ "**.*AdvancedSimulation*" ]
```
Comment (Isammoc) [inline src/test/groovy/func/GatlingRunTaskTest.groovy:36]: Not sure to understand.

Before, it was glob syntax. now it is regex syntax?
If yes:
```suggestion
        project.gatling.includes = [ ".*AdvancedSimulation.*" ]
```
But I guess, it's only the end we want:
```suggestion
        project.gatling.includes = [ ".*AdvancedSimulation" ]
```
As before we had the ".java" extension to accepts, but now, we have the class name.

File: src/test/groovy/func/GatlingRunTaskTest.groovy [hunk 2]
```diff
@@ -43,17 +44,17 @@ class GatlingRunTaskTest extends GatlingUnitSpec {
         given:
         project.gatling.simulations = GatlingPluginExtension.DEFAULT_SIMULATIONS
         and:
-        project.gatlingRun.simulations = { include "**/*AdvancedSimulation*" }
+        project.gatlingRun.includes = [ "**.*AdvancedSimulation*" ]
```
Comment (Isammoc) [inline src/test/groovy/func/GatlingRunTaskTest.groovy:47]: same

File: src/test/groovy/func/GatlingRunTaskTest.groovy (hunk 1 above)
Comment (Isammoc) [reply src/test/groovy/func/GatlingRunTaskTest.groovy:36]: Note, it means this is a breaking change!

File: src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy [hunk 3]
```diff
@@ -327,7 +325,8 @@ class GatlingPluginExtension {
         gatlingVersion = toolVersion
     }
 
-    Closure simulations = DEFAULT_SIMULATIONS
+    List<String> includes = null
+    List<String> excludes = null
```
[the remark this answers is not in the document]
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy:329]: Not sure . I think null is we know the user has set a value.

File: src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy [hunk 4]
```diff
@@ -27,7 +27,10 @@ class GatlingRunTask extends DefaultTask {
     Map environment = [:]
 
     @Internal
-    Closure simulations
+    List<String> includes = null
+
+    @Internal
+    List<String> excludes = null
```
[the remark this answers is not in the document]
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy:33]: So actually, you want to add a feature, right?

File: src/main/groovy/io/gatling/gradle/SimulationFilesUtils.groovy [hunk 5]
```diff
@@ -1,64 +1,47 @@
...
 
-        def scalaSrcDirs = project.sourceSets.gatling.scala.srcDirs.collect { Paths.get(it.absolutePath) }
-        def scalaFiles = getScalaSimulationSources(project, simulations).collect { Paths.get(it.absolutePath) }
+        def allSimulationClasses =  SimulationScanner.scan(dependencies, classDirectories).simulationClasses
 
-        def scalaFQNs = scalaFiles.collect { Path srcFile ->
-            scalaSrcDirs.find { srcFile.startsWith(it) }.relativize(srcFile).join(".") - ".scala"
-        }
+        List<String> includesList = includes? includes : List.of()
+        List<String> excludesList = excludes? excludes : List.of()
```
[the remark this answers is not in the document]
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/SimulationFilesUtils.groovy:25]: Ha, I was looking for this one and couldn't remember where.
Fixing.

File: src/test/groovy/func/GatlingRunTaskTest.groovy (hunk 1 above)
Comment (slandelle) [reply src/test/groovy/func/GatlingRunTaskTest.groovy:36]: Now, it's ant patterns (I think plexus-util also supports regex, but you need to add a prefix. We won't advertise it).

> Note, it means this is a breaking change!

Absolutely. Before the filters were working on source files, now they are working on class names.

File: src/test/groovy/func/GatlingRunTaskTest.groovy [hunk 6]
```diff
@@ -1,17 +1,18 @@
...
+import spock.lang.Ignore
 import spock.lang.Unroll
 
 import static io.gatling.gradle.GatlingPlugin.GATLING_RUN_TASK_NAME
-import static io.gatling.gradle.GatlingPluginExtension.SCALA_SIMULATIONS_DIR
 import static org.apache.commons.io.FileUtils.copyFileToDirectory
 import static org.apache.commons.io.FileUtils.moveFileToDirectory
 
-class GatlingRunTaskTest extends GatlingUnitSpec {
+@Ignore
```
Comment (slandelle) [inline src/test/groovy/func/GatlingRunTaskTest.groovy:14]: @Isammoc I gave up on migrating this test. It needs the bytecode to be compiled but not launch Gatling tests. And the whole class works with configuring the project for each test. Beyond my gradle/spock skills and honestly I don't care.

File: src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy (hunk 4 above)
[the remark this answers is not in the document]
Comment (Isammoc) [reply src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy:33]: I think they were manipulable before (in ant format), weren't they? (if not, safely ignore)

[the remark this answers is not in the document]
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy:33]: My understanding of `@Internal` is that the user was never able to set `simulations` on the task. It's only used internally when using `gatlingRun-XXX`. I'm going to replace it with a single `simulationClass` String.

We'll see later if we need to introduce a feature to let the user set the filters on the task.

OK?

[the remark this answers is not in the document]
Comment (Isammoc) [reply src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy:33]: ok

File: src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy]: Done

File: src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy (hunk 4 above)
[the remark this answers is not in the document]
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy]: done

File: src/main/groovy/io/gatling/gradle/GatlingPlugin.groovy [hunk 7]
```diff
@@ -98,13 +98,7 @@ final class GatlingPlugin implements Plugin<Project> {
...
 
             if (simulationFQN) {
-                simulations = {
-                    include(
-                        "${simulationFQN.replace('.', '/')}.java",
-                        "${simulationFQN.replace('.', '/')}.scala",
-                        "${simulationFQN.replace('.', '/')}.kt"
-                    )
-                }
+                simulationClass = simulationFQN
```
Comment (Isammoc) [inline src/main/groovy/io/gatling/gradle/GatlingPlugin.groovy:101]: So... no condition?

File: src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy [hunk 8]
```diff
@@ -327,7 +325,8 @@ class GatlingPluginExtension {
         gatlingVersion = toolVersion
     }
 
-    Closure simulations = DEFAULT_SIMULATIONS
+    List<String> includes = List.of()
+    List<String> excludes = List.of()
```
Comment (Isammoc) [inline src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy:329]: immutables?
Cannot do a `+=` ?

File: src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy [hunk 9]
```diff
@@ -84,7 +84,11 @@ class GatlingRunTask extends DefaultTask {
         }
     }
 
-    Iterable<String> simulationFilesToFQN() {
-        return SimulationFilesUtils.resolveSimulations(project, this.simulations)
+    List<String> simulationClasses() {
+        def classpath = project.configurations.gatlingRuntimeClasspath.getFiles()
+        def includes = this.simulationClass? List.of(simulationClass) : project.gatling.includes
+        def excludes = this.simulationClass? List.of() : project.gatling.excludes
```
Comment (Isammoc) [inline src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy:90]: nit: format
```suggestion
        def includes = this.simulationClass ? List.of(simulationClass) : project.gatling.includes
        def excludes = this.simulationClass ? List.of() : project.gatling.excludes
```

(this is the ternary operator, not the nullable operator of kotlin)

**Review by Isammoc — APPROVED** [review]

File: src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy (hunk 8 above)
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingPluginExtension.groovy:329]: We'll see if someone needs this. I can't see a use case of this.

File: src/main/groovy/io/gatling/gradle/GatlingPlugin.groovy (hunk 7 above)
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingPlugin.groovy:101]: I don't get it.
This class name is turned into a filter, cf https://github.com/gatling/gatling-gradle-plugin/pull/99/files#diff-4a9f5862b5a900fb63e18d537d80dc750764c83b0b749b0ae9cfb9afa1a8f3b4R89

File: src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy (hunk 9 above)
Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingRunTask.groovy]: Done

File: src/main/groovy/io/gatling/gradle/GatlingPlugin.groovy (hunk 7 above)
Comment (Isammoc) [reply src/main/groovy/io/gatling/gradle/GatlingPlugin.groovy:101]:
```diff
-if (simulationFQN) {
-  simulationClass = simulationFQN
-}
+simulationClass = simulationFQN
```

If null we won't assign it, but it was null... so it won't hurt. Or do I miss something else?

Comment (slandelle) [reply src/main/groovy/io/gatling/gradle/GatlingPlugin.groovy:101]: Ahhh. No, it can't hurt.

**Pull request merged**